File: addon/lib/mode/ofb.cc

#include "ofb.h"
#include <cmath>

namespace cryptian {

namespace mode {

namespace ofb {

bool Ofb::isPaddingRequired() {
    return false;
}

std::vector<char> Cipher::transform(const std::vector<char> chunk) {

    const std::size_t blockSize = _algorithm->getBlockSize();

    std::vector<char> blocks(chunk.size());

    for (size_t i = 0 ; i < chunk.size(); i++) {

        std::vector<char> cipher = _algorithm->encrypt(_register);

        blocks[i] = chunk[i] ^ cipher[0];

        std::copy(_register.begin() + 1, _register.end(), _register.begin());
        _register[blockSize - 1] = cipher[0];

    }

    return blocks;
}


std::vector<char> Decipher::transform(const std::vector<char> chunk) {

    const std::size_t blockSize = _algorithm->getBlockSize();

    std::vector<char> blocks(chunk.size());

    for (size_t i = 0 ; i < chunk.size(); i++) {

        std::vector<char> cipher = _algorithm->encrypt(_register);

        std::copy(_register.begin() + 1, _register.end(), _register.begin());
        _register[blockSize - 1] = cipher[0];

        blocks[i] = chunk[i] ^ cipher[0];

    }

    return blocks;

}

};

};

};

File: test/transform/block/cast-128/ofb.ts

import {expect} from '@jest/globals';

import { default as cryptian, 
    createEncryptStream, createDecryptStream } from "../../../..";

const {algorithm, mode} = cryptian;

import assert from 'assert';

import streamBuffers from 'stream-buffers';

(typeof algorithm.Cast128 === 'function' && typeof mode.ofb === 'object' ? describe : describe.skip) ('cast-128 transform ofb mode', () => {

    const key = Buffer.alloc(16, 0);

    for (let i = 0; i < 16; i++) {
        key[i] = ((i * 2 + 10) % 256);
    }

    const plaintext = Buffer.from(
        '9907d61dcfc9456b237d002ce859b5a1d7bc92747b9cc9d76e' +
        '968442060fa16c9a2f3dd1d2205e19a92c2ad56d12e927224e', 'hex');

    const iv = Buffer.from('8f45b675b98a45ad', 'hex');



    const ciphertext = Buffer.from(
        '0972b74067b62d598246efc5944bb07b0a690e94b481d27e40' +
        '5ff1643a17ea609213d9cc6708e45bdcc115f3cbc2bc58cb0e', 'hex');

    it('should encrypt', () => {
    
        const cast128 = new algorithm.Cast128();
        cast128.setKey(key);

        const cipher = new mode.ofb.Cipher(cast128, iv);

        const transform = createEncryptStream(cipher);
        const buffer = transform.pipe(new streamBuffers.WritableStreamBuffer());

        buffer.on('finish', () => {
            const contents = buffer.getContents();

            expect(contents).toBeTruthy();

            if (contents !== false) {

                assert(ciphertext.equals(contents), 'encrypted plaintext should be equal to ciphertext');

            }
        });

        transform.write(plaintext.slice(0, 22));
        transform.write(plaintext.slice(22, 39));
        transform.end(plaintext.slice(39));
        
    });

    it('should decrypt', () => {

        const cast128 = new algorithm.Cast128();
        cast128.setKey(key);

        const decipher = new mode.ofb.Decipher(cast128, iv);

        const transform = createDecryptStream(decipher);
        const buffer = transform.pipe(new streamBuffers.WritableStreamBuffer());

        buffer.on('finish', () => {
            const contents = buffer.getContents();

            expect(contents).toBeTruthy();

            if (contents !== false) {

                assert(plaintext.equals(contents), 'decrypted ciphertext should be equal to plaintext');

            }
        });

        transform.write(ciphertext.slice(0, 27));
        transform.write(ciphertext.slice(27, 42));
        transform.end(ciphertext.slice(42));
        
    });

    

});

